### ISSUE TITLE
Error on file auto_yus_us.bin

### ISSUE THREAD

**Issue by wandersonareis** [Albeoris/Pulse#19]
I have a problem with this file, auto_yus_us.bin. Using unpack and repack without editing, the game crashes.
I could not figure out the problem because all the other files are edited perfectly.

**Comment by rhadamants** [comment]
Any news about fixing this error?
Only this file is in trouble. The other files work fine.

**Comment by rhadamants** [comment]
> 
> 
> > Any news about fixing this error?
> > Only this file is in trouble. The other files work fine.
> 
> Hello, I also have the same issue. I am translating this game into russian and only this file causes crash.

This auto_yus_us.bin crashes just using the Pulse tool. Without changing anything, just unpack and repack.
But Albeoris still no reply. Very sad for us. This is the only tool I know of for the FF13LR.

**Comment by pinguallyj** [comment]
> > > Any news about fixing this error?
> > > Only this file is in trouble. The other files work fine.
> > 
> > 
> > Hello, I also have the same issue. I am translating this game into russian and only this file causes crash.
> 
> This auto_yus_us.bin crashes just using the Pulse tool. Without changing anything, just unpack and repack.
> But Albeoris still no reply. Very sad for us. This is the only tool I know of for the FF13LR.

I think it has to do with ZTR, the text is uncompressed when I pack it, in the original it is compressed. Do you maybe know how to pack ztr, so the text becomes compressed?

**Comment by rhadamants** [comment]
But all ZTR files are decompressed.
Why just this file would issue to be unpacked?
Pulse does not compress any of the files.

**Comment by pinguallyj** [comment]
> But all ZTR files are decompressed.
> Why just this file would issue to be unpacked?
> Pulse does not compress any of the files.

Maybe it's some weird coding, you know Square enix.
If you look at hex of changed file, all of the text is decompressed.
In original it is totally compressed. Maybe if we can compress it the same way as in original, then the game will work.

**Comment by rhadamants** [comment]
But all the files after Pulse are uncompressed. Why would this file be a problem? The archives of other cities are uncompressed and work.
This is the only file that causes a problem.

**Comment by pinguallyj** [comment]
> But all the files after Pulse are uncompressed. Why would this file be a problem? The archives of other cities are uncompressed and work.
> This is the only file that causes a problem.

https://ibb.co/3dgMLfS
This is auto_yus before we pack ztr
https://ibb.co/HF97Y2p
This is auto_yus after we pack ztr

Now to compare with other files
https://ibb.co/J5LbDpj
This is auto_wil AFTER we pack ztr

So you see, maybe this auto_yus file has another compression method that is not supported by pulse. or somehow it fails to put the text back, because there is different structure or it cant compress it back

**Comment by rhadamants** [comment]
Maybe someone revert this  ZtrFileTextUnpacker and make possible compress ZTR file.

**Comment by pinguallyj** [comment]
> Maybe someone revert this ZtrFileTextUnpacker and make possible compress ZTR file.

Actually, I tested it one more time, if I just pack it without changing, then it is ok.
But as soon as atleast one character changes, then game crashes. This is so weird...

**Comment by rhadamants** [comment]
Just using the file in the tool already crashes. You don't need to edit anything.

**Comment by pinguallyj** [comment]
> Just using the file in the tool already crashes. You don't need to edit anything.

Hm? The tool doesn't crash for me, it does unpack it/pack it,

**Comment by rhadamants** [comment]
Game crash if use this file on Pulse. You do not need to edit anything in the file. Just unpack and repack using this Pulse and the crash game

**Comment by pinguallyj** [comment]
> Game crash if use this file on Pulse. You do not need to edit anything in the file. Just unpack and repack using this Pulse and the crash game

For me if I just unpack/repack the game doesn't crash, only if I change 1 symbol inside ztr.

**Comment by pinguallyj** [comment]
Now it crashes even if it is compressed when I open it with hex. So that is not the problem, I guess.
Interesting thing is, if you put any random file instead of auto_yus then game works. Like if I replace it with auto_wil, it is ok for a game. But as soon as pulse does something ,then it's crash. 
My programming skills are not really good, so I don't know what can cause this...
Maybe it's during wpd packing process?

**Comment by pinguallyj** [comment]
What I also found
https://ibb.co/Gv4ktkt
any other auto file has 12 sets of 00 after header
https://imgbb.com/
auto_yus doesnt have it

**Comment by pinguallyj** [comment]
What I also noticed that after packing file thru pulse, those 00 bytes after header get removed 
https://ibb.co/MZXbkfX
This is original auto_yus and after we pack it, those 00 bytes are gone.
https://ibb.co/MZXbkfX

Still game runs fine, if we don't change anything. But maybe it corrupts the structure of the file somehow?

**Comment by rhadamants** [comment]
You can insert ZTR into WPD manually via Hex. Albeoris needs to look at this. Why only this file causes a problem. When all other files work when using Pulse, it is strange that this file causes a problem.

**Comment by pinguallyj** [comment]
> You can insert ZTR into WPD manually via Hex. Albeoris needs to look at this. Why only this file causes a problem. When all other files work when using Pulse, it is strange that this file causes a problem.

He told me that files are inserted at the end of archive and it can map to a memory which is used by the game. So we need maybe some packer that can put game files in order and implement defragmentation, which will remove gaps in archives. (That’s what Albeoris said)

**Comment by rhadamants** [comment]
You could delete a few lines of text and leave the empty text keys. If you leave the text uncompressed with the same size as the original compressed and test whether the game gives crash. If the problem is the file size uncompressed and play memory buffer.
I don't have the game here to test.

**Comment by pinguallyj** [comment]
> You could delete a few lines of text and leave the empty text keys. If you leave the text uncompressed with the same size as the original compressed and test whether the game gives crash. If the problem is the file size uncompressed and play memory buffer.
> I don't have the game here to test.

Okay I will try it. But I tried placing files from Spanish version or Italian version to English Version which have different sizes, then game runs fine, no crash.

**Comment by rhadamants** [comment]
But the uncompressed file is twice the size of the original archive.

**Comment by pinguallyj** [comment]
> But the uncompressed file is twice the size of the original archive.

Ah yeah you are right, maybe I can also try to remove all lines from Ztr and pack too? I will try when I get home

**Comment by rhadamants** [comment]
Using this [00000001_auto_yus_us.zip](https://github.com/Albeoris/Pulse/files/5778420/00000001_auto_yus_us.zip) not crash.

WPF file [auto_yus_us.zip](https://github.com/Albeoris/Pulse/files/5778421/auto_yus_us.zip).
Game not crash using this WPF smaller than original.

**Comment by pinguallyj** [comment]
Yeah you are right, it works. I wrote him about the issue and that we need compression.

**Comment by pinguallyj** [comment]
He told me there is some CompressZTRContent method instead of FakeCompression, which greatly reduces ztr file size, thought not completely like the original file. WIth this, I determined that maximum file size is 256 kb. If you insert just one more line, it crashes. I used my Russian translation file and for me it is about 1176 lines. After 1177 it crashes.

**Comment by rhadamants** [comment]
256kb for this WPD and ZTR pack or 256kb for ZTR file?
1176 lines for strings files?

**Comment by pinguallyj** [comment]
> 256kb for this WPD and ZTR pack or 256kb for ZTR file?
> 1176 lines for strings files?

256 kb for WPD with ZTR.
https://drive.google.com/file/d/1YerCZryZGS-y5NdhucaQ7l72GGuLam5B/view?usp=sharing
Here is working file.

**Comment by rhadamants** [comment]
Can you send me the method to use CompressZTR?

**Comment by pinguallyj** [comment]
> Can you send me the method to use CompressZTR?

It is already in Pulse, in file ZtrFileKeysPacker, change FakeCompression to CompressZTRContent.
As Albeoris said, this only compresses keys of ztr, as it was needed by Final Fantasy XIII-2 menus to work correctly, as I remember.

**Comment by rhadamants** [comment]
I didn't use this on the FF13-2.
The solution was to use something that already existed and I have the translation to finalize the files for so many years.
Did you make any modifications to use Pulse on the FF13LR or as I only use Pulse on ZTR?

**Comment by pinguallyj** [comment]
> I didn't use this on the FF13-2.
> The solution was to use something that already existed and I have the translation to finalize the files for so many years.
> Did you make any modifications to use Pulse on the FF13LR or as I only use Pulse on ZTR?
I know little about coding and only can read it.
I also use it only for ztr/file extraction and image previews. I did wpd unpacking with your GUI add-on.

**Comment by rhadamants** [comment]
ZtrFileKeysPacker use CompressZTRContent and ZtrFileTextPacker use FakeCompress.
Where should I change?

**Comment by pinguallyj** [comment]
> ZtrFileKeysPacker use CompressZTRContent and ZtrFileTextPacker use FakeCompress.
> Where should I change?

Ah then change in TextPacker

**Comment by rhadamants** [comment]
Did not work. Which line did you change?

**Comment by rhadamants** [comment]
[auto_yus_us.zip](https://github.com/Albeoris/Pulse/files/5782045/auto_yus_us.zip)
258kb. Should I reduce the text to be 256kb?

**Comment by pinguallyj** [comment]
> [auto_yus_us.zip](https://github.com/Albeoris/Pulse/files/5782045/auto_yus_us.zip)
> 258kb. Should I reduce the text to be 256kb?

Yes, otherwise it will crash. I tested with other files like auto_wil and it is the same.
Why auto_yus is special - it has many lines and that is why wdb file is so big 190 kb. that is like more than half of a file.

**Comment by rhadamants** [comment]
For tests game not crash with this yus 258kb. 
But I was unable to verify the texts in the game. why am I using only the yus file.
If you extract this file with compressZTRContent the texts are wrong.

**Comment by rhadamants** [comment]
> 
> 
> > Can you send me the method to use CompressZTR?
> 
> It is already in Pulse, in file ZtrFileKeysPacker, change FakeCompression to CompressZTRContent.
> As Albeoris said, this only compresses keys of ztr, as it was needed by Final Fantasy XIII-2 menus to work correctly, as I remember.

>  // WARNING BUG: Don't use it for text packing or fix BlockOffset first!
        // It should be contains offset if the begining of the entry was mixed with a previous tail.

This warning needs to fix this BlockOffset.
Did you fix this BlockOffset?
Why if you unzip the compressed text appears problems in the strings file.
I did not test in the game if the text gets in trouble.

**Comment by pinguallyj** [comment]
> > > Can you send me the method to use CompressZTR?
> > 
> > 
> > It is already in Pulse, in file ZtrFileKeysPacker, change FakeCompression to CompressZTRContent.
> > As Albeoris said, this only compresses keys of ztr, as it was needed by Final Fantasy XIII-2 menus to work correctly, as I remember.
> 
> > // WARNING BUG: Don't use it for text packing or fix BlockOffset first!
> > // It should be contains offset if the begining of the entry was mixed with a previous tail.
> 
> This warning needs to fix this BlockOffset.
> Did you fix this BlockOffset?
> Why if you unzip the compressed text appears problems in the strings file.
> I did not test in the game if the text gets in trouble.

Yeah, some text gets corrupted with this, because of the tags. But auto yus has no tags and that’s why it works fine, I think. At least for me there is no problem. But don’t pack any other text with it.

**Comment by rhadamants** [comment]
Using the extraction of this compressed auto-yus the string file is missing text characters.

Are you somewhere in the game where I can test this file?
I tested it and it didn't crash but I couldn't find where the texts appear but walking around the city of Yusnan. I have save before facing Snow before opening the door with Sarah's necklace.

**Comment by pinguallyj** [comment]
> Using the extraction of this compressed auto-yus the string file is missing text characters.
> 
> Are you somewhere in the game where I can test this file?
> I tested it and it didn't crash but I couldn't find where the texts appear but walking around the city of Yusnan. I have save before facing Snow before opening the door with Sarah's necklace.

Ah it’s because you also need to change encoding and how Pulse encodes text. For me I changed it from CreateEuro to CreateCyrillic, then text started to show. Maybe you can try to do that. But I don’t remember where is this method should be changed. In visual studio just search for CreateEuro and try to replace it.

**Comment by pinguallyj** [comment]
> > Using the extraction of this compressed auto-yus the string file is missing text characters.
> > Are you somewhere in the game where I can test this file?
> > I tested it and it didn't crash but I couldn't find where the texts appear but walking around the city of Yusnan. I have save before facing Snow before opening the door with Sarah's necklace.
> 
> Ah it’s because you also need to change encoding and how Pulse encodes text. For me I changed it from CreateEuro to CreateCyrillic, then text started to show. Maybe you can try to do that. Because previously before changing I tried to unpack ztr and it was just ###### symbols, is it the same problem? 
But I don’t remember where is this method should be changed. In visual studio just search for CreateEuro and try to replace it.

**Comment by rhadamants** [comment]
I'm using like this:
**var unpacker = new ZtrFileUnpacker(fs, FFXIIITextEncodingFactory.CreateEuro());**
change to?
**var unpacker = new ZtrFileUnpacker(fs, FFXIIITextEncodingFactory.CreateCyrillic());**

This works to extract or pack.
Should I use CreateCyrillic to pack this auto_yus?

**Comment by rhadamants** [comment]
[00000001_auto_yus_us.zip](https://github.com/Albeoris/Pulse/files/5945586/00000001_auto_yus_us.zip)

Unpacked missing characteres:
"The Patron seems like his old self again. I'm gl";
".{End}{End}I hear the savior knocked over the statue.
".{End}{End}Inquisitors from Luxerion are coming here to investigate.

I have doubts if this happens during the game.

Can you send me your auto_yus file that you compressed?

**Comment by pinguallyj** [comment]
Yeah it inserts those random tags during packing because of this compress method, but for my Russian text it worked for some reason. I don’t have that file available right now since I’m at another city. Try change the text around, maybe it will work. Or just use default method to compress but remove some parts of text where lines repeat and also remove those script files in web as well, that can help to shrink the size.
